pkg/dns/dns_test.go

```diff
@@ -15,23 +15,48 @@
 package dns
 
 import (
+	"context"
 	"crypto/tls"
 	"crypto/x509"
 	"io/ioutil"
+	"log"
 	"net"
+	"os"
 	"path"
 	"testing"
 	"time"
 
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	ads "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/protobuf/proto"
 	"github.com/miekg/dns"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/config/memory"
+	"istio.io/istio/pilot/pkg/model"
+	v2 "istio.io/istio/pilot/pkg/proxy/envoy/v2"
+	"istio.io/istio/pilot/pkg/serviceregistry"
+	"istio.io/istio/pilot/pkg/serviceregistry/aggregate"
+	"istio.io/istio/pilot/pkg/serviceregistry/external"
+	"istio.io/istio/pkg/adsc"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/mesh"
+	"istio.io/istio/pkg/config/schema/collections"
 
 	"istio.io/istio/pkg/test/env"
+
+	_ "google.golang.org/grpc/xds/experimental" // To install the xds resolvers and balancers.
```

Comment:
this seems misplaced in the `pkg/dns` package?

Reply:
Not entirely - the DNS client will use the same 'api-based' mode to fetch the Cluster VIP mappings. 

I'll probably split this PR in 2 parts, as well as the tests - one is needed for DNS optimizations, the other is just for grpc.